### ISSUE TITLE
mupdf: very slow opening after crengine profile applied

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#10286]
* KOReader version: 2023.03-20
* Device: Onyx Boox Kon-Tiki 2 (Android 10)

#### Issue
A pdf book became very slow to open after I accidentally applied one of my profiles intended for crengine, while reading that pdf. 

#### Steps to reproduce
Setup a simple crengine profile that changes font face, font size, line spacing, font weight, and kerning. 

Assign it to a gesture with “Quick Menu”. 

Active the quick menu via gesture, tap the profile. 

It will “apply” without complaints (or maybe there were, but I switched off all notifications). 

Now open another book and try to go back. That pdf book will now take a very long time to open, like 1 minute, instead of 3 sec.

What helped was I waited until it finally opened, and reset its document settings. It started to open normally. 

#####`crash.log` (if applicable)
Will try to get the logs later this evening.

**Comment by Frenzie** [comment]
Depending on the PDF and the settings that could be expected, although that does sound surprisingly slow. Changing the font size in a PDF is typically fairly fast but that's no guarantee. Add in a bit of line spacing and you're adding some more time. Most or all of the other settings you mentioned wouldn't carry over. And a PDF that takes 3 seconds to load is probably fairly heavy, so I'd easily expect that to exceed 10 or maybe even 20 seconds. 60 seems a bit much. :-)

**Comment by mergen3107** [comment]
But the problem is that these actions for crengine are in the section for “reflowable documents” in the dispatcher.

If I understand this correctly, this should have absolutely no effect on a mupdf file. 

Also, my current mupdf settings are not involving reflow to be able to have font size and line spacing changeable. 

So I am very confused before I get the logs, why reflowable docs settings even tried to be applied to the fixed layout doc?

**Comment by NiLuJe** [comment]
Wasn't there also the thing where we managed to bundle the application of a bunch of CRe settings in one go, but maybe not for PDFs?

**Comment by hius07** [comment]
Actually only the Font Size setting affects.
For cre documents font size is a font size, but for pdf it is not:
https://github.com/koreader/koreader/blob/d39cc78f9c5013b79f8c2c9e114164e34b1d2361/frontend/ui/data/koptoptions.lua#L9-L10
Applying cre Font Size setting to pdf saves, say, `font_size=22` to doc_settings. Then when opening a document we have:
https://github.com/koreader/koreader/blob/d39cc78f9c5013b79f8c2c9e114164e34b1d2361/frontend/document/pdfdocument.lua#L61-L66
i.e. trying to show a document with font_size=22*22.

**Comment by poire-z** [comment]
Will be hard to change one of the two setting name (migrate docsettings, settings, profiles, gestures...)
So, may be just check for the value in their respective handler?
As it seems cre's readerfont limits its min to 12:
if >= 10, it's cre, pdf font size handler should ignore it
if < 10, it's pdf/kopt, cre font size handler shouldl ignore it?

**Comment by hius07** [comment]
Already done for cre
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/apps/reader/modules/readerfont.lua#L252-L254

**Comment by poire-z** [comment]
That's the check I had found :) But it will _not_ not handle it and just return, it will cap it to 12 and apply 12, so if sending font_size = 1.5 from a PDF font size gesture on a credocument, we should get the CRE document having a font size of 12 ?

**Comment by hius07** [comment]
Applying pdf-font_size action to a cre document has no effect.
It has no immediate effect because it sends `FontSizeUpdate` event that is handled by ReaderView sending the `ReZoom` event that does no action in cre.
doc_setting saves the invalid `copt_font_size` value but it has no effect because cre reads `font_size` value which is intact.

**Comment by hius07** [comment]
The dirty but effective trick would be to add `font_size` check to
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/apps/reader/modules/readerkoptlistener.lua#L74-L79
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/apps/reader/modules/readercoptlistener.lua#L71-L75

**Comment by Frenzie** [comment]
@mergen3107 
> But the problem is that these actions for crengine are in the section for “reflowable documents” in the dispatcher.
> 
> If I understand this correctly, this should have absolutely no effect on a mupdf file.

I wouldn't know otoh without checking but it could. For the most part there would be no effect because there literally can be no effect, not because it's shielded off in case you try to hack the system.

For opening EPUB in MuPDF I made sure that the font-size is carried over though (and if @hius07 says that doesn't happen anymore then I'd be more inclined to consider that a bug really ^_^).

**Comment by Frenzie** [comment]
@hius07 Possibly relevant: https://github.com/koreader/koreader/pull/5282

**Comment by hius07** [comment]
The general structural problem is that if we apply any pdf setting to a cre document (or vice versa), doc_setting will save this setting, for example we can have stupid `copt_zoom_factor`.
Invalid settings are ignored (so the issue is just a garbage in sdr) but `font_size` and `line_spacing` are valid for both formats.

**Comment by hius07** [comment]
Ranges (for reference):

pdf font size
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/ui/data/koptoptions.lua#L10
erroneous comment
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/document/pdfdocument.lua#L62
cre font size
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/ui/data/creoptions.lua#L440-L441
but
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/apps/reader/modules/readerfont.lua#L252-L254

pdf line spacing
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/ui/data/koptoptions.lua#L369
cre line spacing (%)
https://github.com/koreader/koreader/blob/827581ac728330b38853450ff59c3afca00fe488/frontend/ui/data/creoptions.lua#L389-L390

**Comment by mergen3107** [comment]
Well, I think that this fundamental confusion in the code that really needs to be resolved then. 

1) “font size” and “font scale factors” are definitely different things, in my opinion. They should not be handled by the same “font size” variable/parameter/menu in the UI. Font size is absolute, font scale factors is relative, correct?

2) In the UI, it is even more confusing, because it says “REFLOWABLE documents”. That means (from the common sense) the setting must be applied ONLY in a crengine or when mupdf’s reflow is on. So I really think there should be proper checks before the setting can be written to doc settings.

**Comment by Frenzie** [comment]
> “font size” and “font scale factors” are definitely different things, in my opinion. They should not be handled by the same “font size” variable/parameter/menu in the UI. Font size is absolute, font scale factors is relative, correct?

If they're different things they're different things. I'm not saying what's going on, I'm just saying it's completely logical that it might apply. Maybe it does, maybe it doesn't. If applying a setting has an effect that would seem to suggest that it does.

> In the UI, it is even more confusing, because it says “REFLOWABLE documents”. That means (from the common sense) the setting must be applied ONLY in a crengine or when mupdf’s reflow is on. So I really think there should be proper checks before the setting can be written to doc settings.

I disagree. You can find a reference to the English-Dutch dictionary in both English and in Dutch. Nothing implies exclusivity.

**Comment by Frenzie** [comment]
@hius07 If you do change things please keep in mind that MuPDF also supports HTML, EPUB and FB2, and that support has only improved since the version we're currently on. That needn't mean taking care of it right now but designing it so that it's open to it. :-)

**Comment by hius07** [comment]
> MuPDF also supports HTML, EPUB and FB2

And now I got in metadata.fb2.lua:
```Lua
["copt_font_size"] = 1,
["font_size"] = 22,
["kopt_font_size"] = 0.5,
```

**Comment by mergen3107** [comment]
> If applying a setting has an effect that would seem to suggest that it does.

But it doesn’t mean it should. 

> I disagree.
Then it should be consistent with at least itself. 
If these settings apply to both cre/mupdf, then they should have their own third category in dispatcher, along with “reflowable” and “fixed layout”. 

All I am saying is from a user perspective. You developers can fight me all you want, but I think it will be other users as me who will complain about this inconsistency again :D

**Comment by Frenzie** [comment]
I'm speaking from a user perspective. Users can disagree. ;-)

Imo it's a developer perspective to think that which is obviously the same should be neatly separated.

**Comment by Frenzie** [comment]
Basically there are two different things.

1. It makes complete sense that "font size" applies anywhere font size applies.
2. If "font size" currently does something weird, that's a different matter!

**Comment by mergen3107** [comment]
No, I am also a user:
1) I see font size in reflowable docs category. 
2) I expect it to be applied to reflowable docs only. Also because there is a dedicated fixed layout category, and font size is NOT in that category. 

So as a dev-blind user I happened to stumble upon the situation when reflowable setting is applied to fixed-layout doc. And now you guys tell me that for some of those settings “it does what it does”.

**Comment by Frenzie** [comment]
I have no idea what it does right now. I'm saying that's what I think *should* happen.

**Comment by Frenzie** [comment]
That's not speaking as a dev. As a dev my opinion is,eek keep them separated, trying to make it work on both is horribly complex, don't even think about it.

**Comment by mergen3107** [comment]
[crash_mupdf1.log](https://github.com/koreader/koreader/files/11132880/crash_mupdf1.log)
I will comments in a sec...

**Comment by Frenzie** [comment]
Wow, did you repeat the issue that many times? :-)

**Comment by mergen3107** [comment]
I was trying the get verbose dev logs and I don’t think I captured it. Here is what I did. 

1) Enabled verbose logs, restarted KOReader. 
2) Opened that pdf file. 
3) Summoned quick menu via gestures, apply Academy font profile. 
4) Tried to swipe to turn page forward - found out that KOReader is frozen. Wait a bit (roughly 40-60 sec), KOReader turned the page. 
5) Swipe back to previous page - same delay, but turned the page eventually. 
6) While I was waiting, I also tapped at the top, so after finishing page turn it summoned the top menu.

(I wish I had tapped Report a bug button here…)

7) Switched back to a fb2 book.
8) Then switched back to that pdf. 
9) Observed the popup saying the filename is being opened on a blank page background. 
10) Eventually KOReader crashed. 

11) I opened KOReader, and from History reset that pdf’s settings. 
12) It went away from History so I found it in FM again. 
13) Opened it, tried Academy profile again. KOReader since now on crashes after this action after about 10 sec.

What is weird is that it “crashed” so that I saw part of Android top bar with clock and battery percentage. The rest of the screen is white/blank. 

Then after these same 40-60 sec it shows the usual bomb crash details. When I tried to hit “Save Report” button it cleanly (without being in Task Manager) crashed to launcher. 

So I can’t hit the Report a bug button right after applying the font profile anymore :(

I hope the log above contains something useful…

**Comment by mergen3107** [comment]
> Wow, did you repeat the issue that many times? :-)

Yes :D 
See my comment above, I tried many times. I hope it captured attempts in my steps 1-6 :D

**Comment by mergen3107** [comment]
The crash log above is taken somewhere after step 13 and few crashes

**Comment by hius07** [comment]
Out of Memory killed I believe.
No need to apply the full profile, set Font Size is enough to crash.

**Comment by mergen3107** [comment]
Out of memory? Oh, I have 3 GB of RAM O_o

**Comment by Frenzie** [comment]
The limit is realistically probably at most 2 GB (32-bit), though that doesn't make much difference as to what you're saying of course.

**Comment by mergen3107** [comment]
I noticed that AFWall+ was closed as well after KOReader crashes after step 13. So it indeed might be OOM :D

**Comment by NiLuJe** [comment]
Exponentially large font sizes would certainly cause a ridiculously large alloc, so at least that tracks ;).

**Comment by offset-torque** [comment]
>  **mergen3107** “font size” and “font scale factors” are definitely different things, in my opinion.

Another mystery solved. I always thought that PDF font size is some legacy code that defines "size" weirdly for some reason. I never thought that it shows "font scale" because then it would say **scale** instead of **size** I thought. I will add this info to the guide.

On one hand it is better to label UI clearly to give the users correct mental model as **mergen3107** explained above (Font size is absolute, font scale factors is relative).

On the other hand, this feature works as the label says: "it changes the size of the font". So user taps a bigger number or "increase" button until they are satisfied so labeling correctly won't change this behavior.

**In short:** I am sure there are other users that is confused by that UI like me but it doesn't impair the function it seems. Maybe "size" label is even better (less confusing) for the less technical users.

> I noticed that AFWall+ was closed as well after KOReader crashes after step 13

This nearly only happens during an OOM condition on my devices and Android is... let's say "very generous" with its RAM usage, (I am looking at you Google Play Services) so 3 GB is not that much (in the context of Android). Probably GSF takes half of it at boot so KOReader's theoretical limit of 2 GB as **Frenzie** noted, can exceed 3 GB.

**Comment by hius07** [comment]
```
error: malloc of array (54254 x 70752 bytes) failed
./luajit: ./ffi/mupdf.lua:78: could not allocate pixmap: malloc of array (54254 x 70752 bytes) failed (1)
stack traceback:
	[C]: in function 'error'
	./ffi/mupdf.lua:78: in function 'merror'
	./ffi/mupdf.lua:1008: in function 'render_for_kopt'
	./ffi/mupdf.lua:1035: in function 'reflow'
	frontend/document/koptinterface.lua:278: in function 'getCachedContext'
	frontend/document/koptinterface.lua:317: in function 'getPageArea'

```

**Comment by NiLuJe** [comment]
3.6GB \o/.

More imprtantly, does this mean the font size event magically assumes
reflow is enabled even if it's not in this context, or does this require
reflow being enabled first to reproduce?

**Comment by mergen3107** [comment]
> More imprtantly, does this mean the font size event magically assumes reflow is enabled even if it's not in this context, or does this require reflow being enabled first to reproduce?

@NiLuJe This was exactly my concern above: the font size setting is applied regardless of the reflow status. It is “blind”. 

@hius07 Thanks for the size! Now it makes sense. 

@offset-torque Google Services are disabled by default on Onyx Boox, but I guess Android 10 is still RAM hungry :D

**Comment by mergen3107** [comment]
@hius07 ! 
Thank you very much for your time!
Tested quickly - works fine. 

I used the progress bar to track RAM usage by KOReader. Before indeed it was shooting from around 100 MB to 1600 MB. 

Now it only did from 70 to 130. So the setting is ignored completely or the max limit size is applied?
Should I worry to reset document settings if I active the wrong profile again?

**Comment by Frenzie** [comment]
The font size will be ignored completely in both directions because they're separated by range. I forgot to test how that works out for EPUB in MuPDF though. :-)
